lib/sessions.js

```diff
@@ -0,0 +1,144 @@
+    }
+
+    if (sessionPool == null || !(sessionPool instanceof ServerSessionPool)) {
+      throw new Error('ClientSession requires a ServerSessionPool');
+    }
+
+    this.topology = topology;
+    this.sessionPool = sessionPool;
+    this.hasEnded = false;
+    this.serverSession = sessionPool.acquire();
+
+    options = options || {};
+    if (typeof options.initialClusterTime !== 'undefined') {
+      this.clusterTime = options.initialClusterTime;
+    }
+  }
+
+  /**
+   *
+   */
+  endSession(callback) {
+    if (this.hasEnded) {
+      return callback(null, null);
+    }
+
+    // TODO:
+    //   When connected to a sharded cluster the endSessions command
+    //   can be sent to any mongos. When connected to a replica set the
+    //   endSessions command MUST be sent to the primary if the primary
+    //   is available, otherwise it MUST be sent to any available secondary.
+    //   Is it enough to use: ReadPreference.primaryPreferred ?
+
+    this.topology.command(
+      'admin.$cmd',
+      { endSessions: 1, ids: [this.id] },
+      { readPreference: ReadPreference.primaryPreferred },
+      err => {
+        this.hasEnded = true;
+
+        // release the server session back to the pool
+        this.sessionPool.release(this.serverSession);
+
+        if (err) return callback(err, null);
+        callback(null, null);
+      }
+    );
+  }
+}
+
+Object.defineProperty(ClientSession.prototype, 'id', {
+  get: function() {
+    return this.serverSession.id;
+  }
+});
+
+/**
+ *
+ */
+class ServerSession {
+  constructor() {
+    this.id = { id: new Binary(uuidV4(), Binary.SUBTYPE_UUID) };
+    this.lastUse = Date.now();
+  }
+
+  /**
+   *
+   * @param {*} sessionTimeoutMinutes
+   */
+  hasTimedOut(sessionTimeoutMinutes) {
+    const idleTimeMinutes = Math.round(
+      (((Date.now() - this.lastUse) % 86400000) % 3600000) / 60000
+    );
+
+    return idleTimeMinutes > sessionTimeoutMinutes;
+  }
+}
+
+/**
+ *
+ */
+class ServerSessionPool {
+  constructor(topology) {
+    if (topology == null) {
+      throw new Error('ServerSessionPool requires a topology');
+    }
+
+    this.topology = topology;
+    this.sessions = [];
+  }
+
+  /**
+   * @returns {ServerSession}
+   */
+  acquire() {
+    const sessionTimeoutMinutes = this.topology.logicalSessionTimeoutMinutes;
+    while (this.sessions.length) {
+      const session = this.sessions.shift();
+      if (!session.hasTimedOut(sessionTimeoutMinutes)) {
+        return session;
+      }
+    }
+
+    return new ServerSession();
+  }
+
+  /**
+   *
+   * @param {*} session
+   */
+  release(session) {
+    const sessionTimeoutMinutes = this.topology.logicalSessionTimeoutMinutes;
+    while (this.sessions.length) {
+      const session = this.sessions[this.sessions.length - 1];
+      if (session.hasTimedOut(sessionTimeoutMinutes)) {
+        this.sessions.pop();
+      } else {
+        break;
+      }
+    }
+
+    this.sessions.push(session);
```

Comment:
[spec](https://github.com/mongodb/specifications/blob/b1ca52d9b7010ccd40e77d9b5848d33e02a4ab37/source/sessions/driver-sessions.rst#algorithm-to-return-a-serversession-instance-to-the-server-session-pool) seems to say you should only add it to this queue if it's not expiring. I don't think it would have any effect on correctness to add it regardless, since you'll only return it in the future if it's not expired, but it's potentially bad performance-wise?

Reply:
✅